File: packages/executable-stories-pytest/src/executable_stories/_collector.py
"""Thread-safe test case collector.

Accumulates RawTestCase dicts as tests complete, then hands them off
to the JSON writer at session end.
"""

from __future__ import annotations

import threading
from typing import Any


class _Collector:
    """Thread-safe registry for completed test case results."""

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._cases: list[dict[str, Any]] = []
        self._features: dict[str, dict[str, Any]] = {}

    def record(self, test_case: dict[str, Any]) -> None:
        """Append a completed RawTestCase dict."""
        with self._lock:
            self._cases.append(test_case)

    def get_all(self) -> list[dict[str, Any]]:
        """Return all collected test cases."""
        with self._lock:
            return list(self._cases)

    def record_feature(self, feature: dict[str, Any]) -> None:
        """Store a declaration, replacing an earlier one for the same file.

        A re-declaration then reads the way it does in source order.
        """
        with self._lock:
            self._features[feature.get("sourceFile", "")] = feature

    def get_features(self) -> list[dict[str, Any]]:
        """Return all declared features."""
        with self._lock:
            return list(self._features.values())

    def clear(self) -> None:
        """Reset the collector."""
        with self._lock:
            self._cases.clear()
            self._features.clear()


# Module-level singleton
_collector = _Collector()

File: packages/executable-stories-ruby/lib/executable_stories/collector.rb
# frozen_string_literal: true

require_relative "types"

module ExecutableStories
  @mutex = Mutex.new
  @collected = []
  @features = {}
  @order_seq = 0

  module Collector
    module_function

    def record(test_case)
      @mutex ||= Mutex.new
      @collected ||= []
      @mutex.synchronize do
        @collected << test_case
      end
    end

    def all
      @mutex ||= Mutex.new
      @collected ||= []
      @mutex.synchronize do
        @collected.dup
      end
    end

    # Store a declaration, replacing an earlier one for the same file so a
    # re-declaration reads the way it does in source order.
    def record_feature(feature)
      @mutex ||= Mutex.new
      @features ||= {}
      @mutex.synchronize do
        @features[feature.source_file] = feature
      end
    end

    def all_features
      @mutex ||= Mutex.new
      @features ||= {}
      @mutex.synchronize do
        @features.values
      end
    end

    def next_order
      @mutex ||= Mutex.new
      @order_seq ||= 0
      @mutex.synchronize do
        n = @order_seq
        @order_seq += 1
        n
      end
    end

    def reset
      @mutex ||= Mutex.new
      @mutex.synchronize do
        @collected = []
        @features = {}
        @order_seq = 0
      end
    end
  end
end
